Failed to start with Spring AOT cache on

### <!-- Component -->

<!-- Clients- -->

### <!-- Affected version -->

<!-- -8.8 -->

### <!-- Severity -->

<!-- Medium- -->

### <!-- Likelihood -->

<!-- High_ -->

### Description

Failure to start the my Spring Boot 4 application using the Camunda Spring Boot Starter application using the Spring AOT cache.



### Steps to reproduce

1. Build a Spring Boot application using the camunda-spring-boot-4-starter library using spring-boot-maven-plugin and add as execution:
```xml
<goals>
  <goal>process-aot</goal>
</goals>
```
2. After the application is built, start it with "-Dspring.aot.enabled=true"

### Current behavior

In my Spring Boot 4 (using 4.0.3) application I use the Camunda Spring Boot Starter library compatible for Spring Boot 4:
```xml
<dependency>
      <groupId>io.camunda</groupId>
      <artifactId>camunda-spring-boot-4-starter</artifactId>
      <version>8.8.17</version>
</dependency>
```
I build my application generating the Spring AOT cache. When starting the application with "-Dspring.aot.enabled=true", it fails to bootstrap due to:
```
ERROR o.s.boot.SpringApplication - Application run failed 
org.springframework.context.ApplicationContextException: Failed to start bean 'camundaLifecycleEventProducer'
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:423)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:409)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:613)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:379)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:306)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:1013)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:624)
	at org.springframework.boot.web.server.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:756)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:445)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1365)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1354)
	at org.epo.pgp.appeal.Application.main(Application.java:15)
Caused by: java.lang.NullPointerException: Cannot invoke "io.camunda.zeebe.spring.client.annotation.customizer.ZeebeWorkerValueCustomizer.customize(io.camunda.zeebe.spring.client.annotation.value.ZeebeWorkerValue)" because "this.customizer" is null
	at io.camunda.client.spring.annotation.customizer.JobWorkerValueCustomizerCompat.customize(JobWorkerValueCustomizerCompat.java:33)
	at io.camunda.client.spring.annotation.processor.JobWorkerAnnotationProcessor.lambda$start$1(JobWorkerAnnotationProcessor.java:89)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
	at io.camunda.client.spring.annotation.processor.JobWorkerAnnotationProcessor.lambda$start$2(JobWorkerAnnotationProcessor.java:88)
	at java.base/java.util.stream.ReferencePipeline$15$1.accept(ReferencePipeline.java:576)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1716)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:632)
	at io.camunda.client.spring.annotation.processor.JobWorkerAnnotationProcessor.start(JobWorkerAnnotationProcessor.java:91)
	at io.camunda.client.spring.annotation.processor.AbstractCamundaAnnotationProcessor.onStart(AbstractCamundaAnnotationProcessor.java:59)
	at io.camunda.client.spring.event.CamundaClientEventListener.lambda$handleStart$0(CamundaClientEventListener.java:36)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at io.camunda.client.spring.event.CamundaClientEventListener.handleStart(CamundaClientEventListener.java:35)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.doInvoke(ApplicationListenerMethodAdapter.java:375)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.processEvent(ApplicationListenerMethodAdapter.java:252)
	at org.springframework.context.event.ApplicationListenerMethodAdapter.onApplicationEvent(ApplicationListenerMethodAdapter.java:171)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:180)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:173)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:151)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:448)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:381)
	at io.camunda.client.spring.event.CamundaLifecycleEventProducer.start(CamundaLifecycleEventProducer.java:38)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:420)
	... 14 common frames omitted
```

### Expected behavior

The application should start using the library.

### Workaround

Start the application with "-Dspring.aot.enabled=false"

### Environment

SM

### Version

<dependency>
      <groupId>io.camunda</groupId>
      <artifactId>camunda-spring-boot-4-starter</artifactId>
      <version>8.8.17</version>
</dependency>

### Rootcause

_No response_

### Solution ideas

_No response_

### Dev -> QA handover

- Resources:
- Versions to validate:
- Release version (in which version this feature will be released):


### Links

_No response_

### Automated Test Impact

_No response_

--- a/clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/annotation/customizer/LegacyJobWorkerValueCustomizerBeanDefinitionRegistryPostProcessor.java
+++ b/clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/annotation/customizer/LegacyJobWorkerValueCustomizerBeanDefinitionRegistryPostProcessor.java
@@ -20,7 +20,6 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.BeansException;
 import org.springframework.beans.factory.config.BeanDefinition;
-import org.springframework.beans.factory.support.BeanDefinitionBuilder;
 import org.springframework.beans.factory.support.BeanDefinitionRegistry;
 import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor;
 
@@ -42,11 +41,6 @@ public class LegacyJobWorkerValueCustomizerBeanDefinitionRegistryPostProcessor
         LOG.warn(
             "Bean '{}' is implementing deprecated interface ZeebeWorkerValueCustomizer, please migrate to JobWorkerValueCustomizer",
             beanDefinitionName);
-        final BeanDefinitionBuilder beanDefinitionBuilder =
-            BeanDefinitionBuilder.genericBeanDefinition(JobWorkerValueCustomizerCompat.class)
-                .addConstructorArgReference(beanDefinitionName);
-        registry.registerBeanDefinition(
-            beanDefinitionName + "_Compat", beanDefinitionBuilder.getBeanDefinition());
       }
     }
   }

--- a/clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/configuration/AnnotationProcessorConfiguration.java
+++ b/clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/configuration/AnnotationProcessorConfiguration.java
@@ -18,11 +18,15 @@ package io.camunda.client.spring.configuration;
 import io.camunda.client.annotation.customizer.JobWorkerValueCustomizer;
 import io.camunda.client.jobhandling.JobWorkerManager;
 import io.camunda.client.lifecycle.CamundaClientLifecycleAware;
+import io.camunda.client.spring.annotation.customizer.JobWorkerValueCustomizerCompat;
 import io.camunda.client.spring.annotation.processor.DeploymentAnnotationProcessor;
 import io.camunda.client.spring.annotation.processor.JobWorkerAnnotationProcessor;
 import io.camunda.client.spring.event.CamundaClientEventListener;
+import io.camunda.zeebe.spring.client.annotation.customizer.ZeebeWorkerValueCustomizer;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.context.ApplicationEventPublisher;
 import org.springframework.context.annotation.Bean;
@@ -45,7 +49,15 @@ public class AnnotationProcessorConfiguration {
   @Bean
   public JobWorkerAnnotationProcessor jobWorkerPostProcessor(
       final JobWorkerManager jobWorkerManager,
-      final List<JobWorkerValueCustomizer> jobWorkerValueCustomizers) {
-    return new JobWorkerAnnotationProcessor(jobWorkerManager, jobWorkerValueCustomizers);
+      final List<JobWorkerValueCustomizer> jobWorkerValueCustomizers,
+      @Autowired(required = false) final List<ZeebeWorkerValueCustomizer> legacyCustomizers) {
+    final List<JobWorkerValueCustomizer> allCustomizers =
+        new ArrayList<>(jobWorkerValueCustomizers);
+    if (legacyCustomizers != null) {
+      legacyCustomizers.stream()
+          .map(JobWorkerValueCustomizerCompat::new)
+          .forEach(allCustomizers::add);
+    }
+    return new JobWorkerAnnotationProcessor(jobWorkerManager, allCustomizers);
   }
 }

--- a/clients/camunda-spring-boot-starter/src/test/java/io/camunda/client/spring/annotation/customizer/JobWorkerValueCustomizerCompatTest.java
+++ b/clients/camunda-spring-boot-starter/src/test/java/io/camunda/client/spring/annotation/customizer/JobWorkerValueCustomizerCompatTest.java
@@ -17,12 +17,9 @@ package io.camunda.client.spring.annotation.customizer;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-import io.camunda.client.annotation.customizer.JobWorkerValueCustomizer;
 import io.camunda.client.spring.configuration.CamundaBeanPostProcessorConfiguration;
-import java.util.Set;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.boot.test.system.CapturedOutput;
 import org.springframework.boot.test.system.OutputCaptureExtension;
@@ -34,17 +31,9 @@ import org.springframework.boot.test.system.OutputCaptureExtension;
     })
 @ExtendWith(OutputCaptureExtension.class)
 public class JobWorkerValueCustomizerCompatTest {
-  @Autowired Set<JobWorkerValueCustomizer> jobWorkerValueCustomizers;
 
   @Test
-  void shouldPostProcessBean(final CapturedOutput capturedOutput) {
-    assertThat(jobWorkerValueCustomizers).hasSize(1);
-    final JobWorkerValueCustomizer jobWorkerValueCustomizer =
-        jobWorkerValueCustomizers.iterator().next();
-    assertThat(jobWorkerValueCustomizer).isInstanceOf(JobWorkerValueCustomizerCompat.class);
-    final JobWorkerValueCustomizerCompat customizer =
-        (JobWorkerValueCustomizerCompat) jobWorkerValueCustomizer;
-    assertThat(customizer.getCustomizer()).isNotNull();
+  void shouldLogDeprecationWarningForLegacyBean(final CapturedOutput capturedOutput) {
     assertThat(capturedOutput)
         .contains(
             "Bean 'testZeebeWorkerValueCustomizer' is implementing deprecated interface ZeebeWorkerValueCustomizer, please migrate to JobWorkerValueCustomizer");

--- a/clients/camunda-spring-boot-starter/src/test/java/io/camunda/client/spring/config/LegacyZeebeWorkerCustomizerAotCompatibilityTest.java
+++ b/clients/camunda-spring-boot-starter/src/test/java/io/camunda/client/spring/config/LegacyZeebeWorkerCustomizerAotCompatibilityTest.java
@@ -1,0 +1,103 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.client.spring.config;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+import io.camunda.client.CamundaClient;
+import io.camunda.client.annotation.JobWorker;
+import io.camunda.client.api.response.ActivatedJob;
+import io.camunda.client.jobhandling.JobWorkerManager;
+import io.camunda.client.spring.configuration.AnnotationProcessorConfiguration;
+import io.camunda.client.spring.event.CamundaClientCreatedSpringEvent;
+import io.camunda.client.spring.event.CamundaClientEventListener;
+import io.camunda.zeebe.spring.client.annotation.customizer.ZeebeWorkerValueCustomizer;
+import io.camunda.zeebe.spring.client.annotation.value.ZeebeWorkerValue;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.test.context.bean.override.mockito.MockitoBean;
+
+/**
+ * Regression test for https://github.com/camunda/camunda/issues/49131.
+ *
+ * <p>Verifies that a legacy {@link ZeebeWorkerValueCustomizer} bean is applied to job workers even
+ * when {@code LegacyJobWorkerValueCustomizerBeanDefinitionRegistryPostProcessor} does not register
+ * the compat wrapper as a bean — which is the scenario that occurs with Spring AOT, where the
+ * dynamic {@code BeanDefinitionRegistryPostProcessor} bean registration produces a null-customizer
+ * {@code JobWorkerValueCustomizerCompat} due to an unresolved {@code RuntimeBeanReference}.
+ */
+@SpringBootTest(
+    classes = {
+      AnnotationProcessorConfiguration.class,
+      LegacyZeebeWorkerCustomizerAotCompatibilityTest.TestConfig.class
+    })
+class LegacyZeebeWorkerCustomizerAotCompatibilityTest {
+
+  // No CamundaBeanPostProcessorConfiguration — simulates the AOT scenario where the
+  // BeanDefinitionRegistryPostProcessor does not register compat beans.
+  @MockitoBean JobWorkerManager jobWorkerManager;
+  @MockitoBean CamundaClient camundaClient;
+
+  @Autowired CamundaClientEventListener camundaClientEventListener;
+  @Autowired TestConfig.TrackingZeebeCustomizer trackingCustomizer;
+
+  @Test
+  void shouldApplyLegacyZeebeWorkerValueCustomizerWhenPostProcessorIsAbsent() {
+    // when
+    camundaClientEventListener.handleStart(
+        new CamundaClientCreatedSpringEvent(this, mock(CamundaClient.class)));
+
+    // then — the legacy customizer must have been invoked via the compat wrapper,
+    // proving AnnotationProcessorConfiguration handles it directly
+    assertThat(trackingCustomizer.wasCalled()).isTrue();
+  }
+
+  @Configuration
+  static class TestConfig {
+
+    @Bean
+    public TrackingZeebeCustomizer legacyZeebeCustomizer() {
+      return new TrackingZeebeCustomizer();
+    }
+
+    @Bean
+    public WorkerBean workerBean() {
+      return new WorkerBean();
+    }
+
+    static class TrackingZeebeCustomizer implements ZeebeWorkerValueCustomizer {
+      private boolean called = false;
+
+      @Override
+      public void customize(final ZeebeWorkerValue worker) {
+        called = true;
+      }
+
+      public boolean wasCalled() {
+        return called;
+      }
+    }
+
+    static class WorkerBean {
+      @JobWorker(type = "test-type")
+      public void handle(final ActivatedJob job) {}
+    }
+  }
+}

